core/templates/expressions/expression-syntax-tree.service.ts

```diff
@@ -16,417 +16,447 @@
  * @fileoverview Expression syntax tree service.
  */
 
-require('expressions/expression-parser.service.ts');
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
 
-angular.module('oppia').factory('ExpressionSyntaxTreeService', [
-  'ExpressionParserService', 'PARAMETER_TYPES',
-  function(ExpressionParserService, PARAMETER_TYPES) {
-    // Exceptions that can be thrown from the evaluation of expressions.
-    const ExpressionError = function() {};
-    ExpressionError.prototype = new Error();
-    ExpressionError.prototype.constructor = ExpressionError;
+import { AppConstants } from 'app.constants';
+import { ExpressionParserService } from
+  'expressions/expression-parser.service.ts';
 
-    var ExprUndefinedVarError = function(varname, envs) {
-      this.varname = varname;
-      this.envs = envs;
-    };
-    ExprUndefinedVarError.prototype = new ExpressionError();
-    ExprUndefinedVarError.prototype.constructor = ExprUndefinedVarError;
-    ExprUndefinedVarError.prototype.name = 'ExprUndefinedVarError';
-    ExprUndefinedVarError.prototype.toString = function() {
-      return this.name + ': ' + this.varname + ' not found in ' + this.envs;
-    };
+// Exceptions that can be thrown from the evaluation of expressions.
+const ExpressionError = function(): void {};
```

Comment:
Please make these all classes. Our previous code tried to emulate classes by manually changing prototypes, but now that we are migrating to Angular 8 which uses typescript, that's no longer necessary. As part of the migration, we should be changing these to classes like so:

```typescript
public class ExpressionError {
}

public class ExprUndefinedVarError {
  constructor(public varname: string, public envs: any[]);

  toString(): string {
    // ...
  }
}

// etc...
```

Reply:
@brianrodri Are you sure we can declare nested classes? It doesn't work and not even in TS playground.

Reply:
They shouldn't be nested, they should be top-level.